frontend/src/citizen-frontend/calendar/CalendarGridView.tsx

```diff
@@ -0,0 +1,172 @@
+// SPDX-FileCopyrightText: 2017-2021 City of Espoo
+//
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+import React from 'react'
+import _ from 'lodash'
+import LocalDate from 'lib-common/local-date'
+import { DailyReservationData } from './api'
+import styled from 'styled-components'
+import { useTranslation } from '../localization'
+import colors from 'lib-customizations/common'
+import { WeekProps } from './WeekElem'
+import { H1 } from 'lib-components/typography'
+import { defaultMargins } from 'lib-components/white-space'
+import { formatDate } from 'lib-common/date'
+
+export interface Props {
+  dailyReservations: DailyReservationData[]
+  onCreateReservationClicked: () => void
+  selectDate: (date: LocalDate) => void
+}
+
+export default React.memo(function CalendarListView({
+  dailyReservations,
+  selectDate
+}: Props) {
+  const i18n = useTranslation()
+  const weeklyData = dailyReservations.reduce((weekly, daily) => {
+    const last = _.last(weekly)
+    if (last === undefined || daily.date.getIsoWeek() !== last.weekNumber) {
+      return [
+        ...weekly,
+        { weekNumber: daily.date.getIsoWeek(), dailyReservations: [daily] }
+      ]
+    } else {
+      return [
+        ..._.dropRight(weekly),
+        {
+          ...last,
+          dailyReservations: [...last.dailyReservations, daily]
+        }
+      ]
+    }
+  }, [] as WeekProps[])
```

Comment:
You could just do this higher up the component tree so you wouldn't have to do the same processing twice.